### PR TITLE
Feature/add tests in the classes of encoder field

### PR DESCRIPTION
**Pull request by PauloMiranda98**
This pull request aims to add test units in the classes of the `encoder.field` package.

### CODE REVIEW

File: src/test/java/com/iabtcf/encoder/field/BooleanEncoderTest.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        Assert.assertFalse(BooleanEncoder.getInstance().decode(VALUE_0));
+    }
+
+    @Test
+    public void whenValueIsOneShouldReturnTrue() {
+        Assert.assertTrue(BooleanEncoder.getInstance().decode(VALUE_1));
+    }
+
+    @Test
+    public void whenValueIsInvalidShouldReturnFalse() {
```
Comment (a-s-rose) [inline src/test/java/com/iabtcf/encoder/field/BooleanEncoderTest.java]: Hi Paulo,
this both lines are actually two test cases, please use separate test cases Empty, invalid). It's easier to find errors if the test covers only one assertion.
Also consider testing null parameter.
Greets
Antje

File: src/test/java/com/iabtcf/encoder/field/DateEncoderTest.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    private static final String EMPTY_VALUE = "";
+
+    @Test
+    public void whenValueIsValidShouldReturnExpectedDate() {
+        Assert.assertEquals(EXPECTED_RETURN, DateEncoder.getInstance().decode(VALID_VALUE));
+    }
+
+    @Test
+    public void whenValueIsInvalidShouldReturnNull() {
+        Assert.assertNull(DateEncoder.getInstance().decode(INVALID_VALUE));
```
Comment (a-s-rose) [inline src/test/java/com/iabtcf/encoder/field/DateEncoderTest.java]: See my previous comment for BooleanEncoderTest. Same here.

File: src/test/java/com/iabtcf/encoder/field/FixedVectorEncoderTest.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    @Test
+    public void whenValueIsValidThenShouldReturnExpectedSortedVector() {
+        Assert.assertEquals(EXPECTED_RETURN, FixedVectorEncoder.getInstance().decode(VALID_VALUE));
+    }
+
+    @Test
+    public void whenValueIsInvalidThenShouldReturnNull() {
+        Assert.assertNull(FixedVectorEncoder.getInstance().decode(INVALID_VALUE));
+    }
+
```
Comment (a-s-rose) [inline src/test/java/com/iabtcf/encoder/field/FixedVectorEncoderTest.java]: How does the class under test behave if you fill a string not containing numeric values, say for example "abc102(&$_-"?

File: src/test/java/com/iabtcf/encoder/field/IntEncoderTest.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+    @Test
+    public void whenValueIsValidThenShouldReturnExpectedValue() {
+        Assert.assertEquals(EXPECTED_VALUE_1, IntEncoder.getInstance().decode(VALUE_1));
+        Assert.assertEquals(EXPECTED_VALUE_10, IntEncoder.getInstance().decode(VALUE_10));
+    }
+
+    @Test
+    public void whenValueIsInvalidShouldReturnFalse() {
+        Assert.assertNull(IntEncoder.getInstance().decode(EMPTY_VALUE));
```
Comment (a-s-rose) [inline src/test/java/com/iabtcf/encoder/field/IntEncoderTest.java]: See my previous comment for BooleanEncoderTest. Same here.

File: src/test/java/com/iabtcf/encoder/field/LangEncoderTest.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+    @Test
+    public void whenValueIsValidThenShouldReturnExpectedValue() {
+        Assert.assertEquals(EXPECTED_VALUE_PT, LangEncoder.getInstance().decode(VALUE_PT));
+        Assert.assertEquals(EXPECTED_VALUE_EN, LangEncoder.getInstance().decode(VALUE_EN));
+    }
+
+    @Test
+    public void whenValueIsInvalidShouldReturnFalse() {
+        Assert.assertNull(IntEncoder.getInstance().decode(EMPTY_VALUE));
```
Comment (a-s-rose) [inline src/test/java/com/iabtcf/encoder/field/LangEncoderTest.java]: See my previous comment for BooleanEncoderTest. Same here.

File: src/test/java/com/iabtcf/encoder/field/BooleanEncoderTest.java (hunk 1 above)
Comment (PauloMiranda98) [reply src/test/java/com/iabtcf/encoder/field/BooleanEncoderTest.java]: The suggested change was made.

File: src/test/java/com/iabtcf/encoder/field/DateEncoderTest.java (hunk 2 above)
Comment (PauloMiranda98) [reply src/test/java/com/iabtcf/encoder/field/DateEncoderTest.java]: The suggested change was made.

File: src/test/java/com/iabtcf/encoder/field/IntEncoderTest.java (hunk 4 above)
Comment (PauloMiranda98) [reply src/test/java/com/iabtcf/encoder/field/IntEncoderTest.java]: The suggested change was made.

File: src/test/java/com/iabtcf/encoder/field/LangEncoderTest.java (hunk 5 above)
Comment (PauloMiranda98) [reply src/test/java/com/iabtcf/encoder/field/LangEncoderTest.java]: The suggested change was made.

File: src/test/java/com/iabtcf/encoder/field/FixedVectorEncoderTest.java (hunk 3 above)
Comment (PauloMiranda98) [reply src/test/java/com/iabtcf/encoder/field/FixedVectorEncoderTest.java]: The suggested change was made.

**Comment by PauloMiranda98** [comment]
I finished testing this Pull Request. In my local machine I finished the tests of package `encoder.segment` but I need the feedback of this PR and that it be merged before push another PR.

File: src/test/java/com/iabtcf/encoder/field/FieldEncoderMapTest.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        Assert.assertEquals(FIELD_ENCODER_MAP.numCustomPurposes, getValueOfFieldMap(Fields.numCustomPurposes));
+        Assert.assertEquals(FIELD_ENCODER_MAP.publisherCustomConsents,
+                getValueOfFieldMap(Fields.publisherCustomConsents));
+        Assert.assertEquals(FIELD_ENCODER_MAP.publisherCustomLegitimateInterest,
+                getValueOfFieldMap(Fields.publisherCustomLegitimateInterest));
+    }
+
+    @Test
+    public void whenMapKeyIsInvalidThenTheValueShouldIsNull() {
+        Assert.assertNull(getValueOfFieldMap(INVALID_KEY));
```
Comment (chrispaterson) [inline src/test/java/com/iabtcf/encoder/field/FieldEncoderMapTest.java]: Is this not the default language feature?  If it is then I don't think we need to test it...

File: src/test/java/com/iabtcf/encoder/field/IntEncoderTest.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+    @Test
+    public void whenValueIsValidThenShouldReturnExpectedValue() {
+        Assert.assertEquals(EXPECTED_VALUE_1, IntEncoder.getInstance().decode(VALUE_1));
+        Assert.assertEquals(EXPECTED_VALUE_10, IntEncoder.getInstance().decode(VALUE_10));
+    }
+
+    @Test
+    public void whenValueIsInvalidShouldReturnNull() {
+        Assert.assertNull(IntEncoder.getInstance().decode(INVALID_VALUE));
```
Comment (chrispaterson) [inline src/test/java/com/iabtcf/encoder/field/IntEncoderTest.java]: idem (if language feature – don't test)

File: src/main/java/com/iabtcf/encoder/field/BooleanEncoder.java [hunk 8]
```diff
@@ -11,10 +15,25 @@ private BooleanEncoder() {
...
         return instance;
     }
-    public final Boolean decode(String value){
-        return value.equals("1");
+
+    private static final Logger logger = LogManager.getLogger(IntEncoder.class);
+
+    public final Boolean decode(String value) {
+        try {
+            if (value.equals(TRUE)) {
```
[the remark this answers is not in the document]
Comment (PauloMiranda98) [reply src/main/java/com/iabtcf/encoder/field/BooleanEncoder.java]: The idea is to identify the difference between "0", "1", null and invalid values ("ABC", "$ 231", "", among others). Because it is important to detect whether the string is invalid.

File: src/main/java/com/iabtcf/encoder/field/BooleanEncoder.java [hunk 9]
```diff
@@ -11,10 +15,25 @@ private BooleanEncoder() {
...
+                return true;
+            }
+            if (value.equals(FALSE)) {
+                return false;
+            }
+            return null;
```
[the remark this answers is not in the document]
Comment (PauloMiranda98) [reply src/main/java/com/iabtcf/encoder/field/BooleanEncoder.java]: But how do we detect if the String is invalid or not?

File: src/main/java/com/iabtcf/model/PurposeRestrictionVector.java [hunk 10]
```diff
@@ -124,4 +124,19 @@ public Boolean isValid() {
     public int getNumRestrictions() {
         return this.map.size();
     }
+
+    @Override
+    public boolean equals(Object o) {
```
[the remark this answers is not in the document]
Comment (PauloMiranda98) [reply src/main/java/com/iabtcf/model/PurposeRestrictionVector.java]: I'm going to switch to lombok notation

File: src/main/java/com/iabtcf/model/SortedVector.java [hunk 11]
```diff
@@ -36,4 +35,17 @@ public SortedSet getSet() {
     public void setSet(SortedSet set) {
         this.set = set;
     }
+
+    @Override
+    public boolean equals(Object o) {
```
[the remark this answers is not in the document]
Comment (PauloMiranda98) [reply src/main/java/com/iabtcf/model/SortedVector.java]: I'm going to switch to lombok notation.

File: src/main/java/com/iabtcf/encoder/field/BooleanEncoder.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (chrispaterson) [reply src/main/java/com/iabtcf/encoder/field/BooleanEncoder.java]: If we've made it all the way down to this decoder, we should know that it was invalid because the base64Url decoder checks to make sure it can be decoded into a value:
https://github.com/InteractiveAdvertisingBureau/iabtcf-java/blob/master/src/main/java/com/iabtcf/encoder/Base64Url.java#L35

File: src/test/java/com/iabtcf/encoder/field/FieldEncoderMapTest.java (hunk 6 above)
Comment (PauloMiranda98) [reply src/test/java/com/iabtcf/encoder/field/FieldEncoderMapTest.java]: It's true, I will remove.

File: src/main/java/com/iabtcf/encoder/field/BooleanEncoder.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (chrispaterson) [reply src/main/java/com/iabtcf/encoder/field/BooleanEncoder.java]: If anything, this should use a regex match for a "1" or a "0" and throw an error if it is not either of those because it's an error case that should blow up the whole decode.

File: src/test/java/com/iabtcf/encoder/field/FixedVectorEncoderTest.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    private static final String EMPTY_VALUE = "";
+    private static final SortedVector EMPTY_RETURN = new SortedVector();
+
+    private static final String INVALID_VALUE = "invalid_value";
+    private static final String NULL_VALUE = null;
+
+    @Before
+    public void before() {
+        for (int i = 0; i < VALID_VALUE.length(); i++) {
+            if (VALID_VALUE.charAt(i) == '1')
```
[the remark this answers is not in the document]
Comment (PauloMiranda98) [reply src/test/java/com/iabtcf/encoder/field/FixedVectorEncoderTest.java]: I added the brackets.

File: src/test/java/com/iabtcf/encoder/field/IntEncoderTest.java (hunk 7 above)
Comment (PauloMiranda98) [reply src/test/java/com/iabtcf/encoder/field/IntEncoderTest.java]: But if in the future the implementation of IntDecoder changes? We need to ensure that it continues to return null regardless of how it is implemented.

File: src/main/java/com/iabtcf/encoder/field/BooleanEncoder.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (PauloMiranda98) [reply src/main/java/com/iabtcf/encoder/field/BooleanEncoder.java]: If I no longer use the null as a return, the following tests will no longer pass:
- **BooleanEncoderTest**
whenValueIsEmptyShouldReturnNull
whenValueIsInvalidShouldReturnNull
whenValueIsNullShouldReturnNull

- **FixedVectorEncoderTest**
whenValueIsInvalidThenShouldReturnNull

So do I delete them?

[the remark this answers is not in the document]
Comment (chrispaterson) [reply src/main/java/com/iabtcf/encoder/field/BooleanEncoder.java]: I really believe we should be throwing an error instead of returning `null` because null is really a terminal case.  The values should have been checked further upstream before an encoding was to occur.

File: src/test/java/com/iabtcf/encoder/field/IntEncoderTest.java (hunk 7 above)
Comment (chrispaterson) [reply src/test/java/com/iabtcf/encoder/field/IntEncoderTest.java]: I'm not sure I follow, I don't think it should ever return `null`.

File: src/main/java/com/iabtcf/encoder/field/BooleanEncoder.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (chrispaterson) [reply src/main/java/com/iabtcf/encoder/field/BooleanEncoder.java]: Otherwise you'll have to check for `null` and do something with it, which I believe will end up being halting execution.

[the remark this answers is not in the document]
Comment (a-s-rose) [reply src/main/java/com/iabtcf/encoder/field/BooleanEncoder.java]: You could use a Precondition check (guava Library for example) for the parameter at the beginning of the method (which will actually throw a NullPointerException in case the parameter is null). The calling code on the other hand has to make sure no nulls are passed to the method.
Your tests could then look like that (pseudo code):
`
@Test(expected = NullPointerException.class)
public void testNull() {
    myBooleanDecoder.decode(null);
}
`

[the remark this answers is not in the document]
Comment (PauloMiranda98) [reply src/main/java/com/iabtcf/encoder/field/BooleanEncoder.java]: I will change this.

File: src/test/java/com/iabtcf/encoder/field/IntEncoderTest.java (hunk 7 above)
Comment (PauloMiranda98) [reply src/test/java/com/iabtcf/encoder/field/IntEncoderTest.java]: I will change this.

File: src/main/java/com/iabtcf/encoder/field/BooleanEncoder.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (PauloMiranda98) [reply src/main/java/com/iabtcf/encoder/field/BooleanEncoder.java]: I will change this.

File: src/test/java/com/iabtcf/encoder/field/BooleanEncoderTest.java [hunk 13]
```diff
@@ -0,0 +1,19 @@
+package com.iabtcf.encoder.field;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+public class BooleanEncoderTest {
+    private static final String VALUE_0 = "0";
```
Comment (macfol) [inline src/test/java/com/iabtcf/encoder/field/BooleanEncoderTest.java]: consider TRUE, FALSE for better readability

**Comment by laktech** [comment]
Not clear how relevant this PR this anymore. It's been open for sometime now and the 'develop' branch has evolved beyond this.
